src/lib/utils/lib.ts

```diff
@@ -52,3 +55,109 @@ export const getTransactionReceipt = async (
   typeof val !== 'undefined' && val !== null
+
+export const getBlockRangesForL1Block = async ({
+  targetL1BlockNumber,
+  provider,
+}: {
+  targetL1BlockNumber: number
+  provider: JsonRpcProvider
+}) => {
+  const arbitrumProvider = new ArbitrumProvider(provider)
+  const currentArbBlock = await arbitrumProvider.getBlockNumber()
+
+  const arbitrumChainId = (await arbitrumProvider.getNetwork()).chainId
+  const { nitroGenesisBlock } = l2Networks[arbitrumChainId]
+
+  // Define the starting point to be closer to the current block for efficiency.
+  let startArbBlock = Math.floor(currentArbBlock * 0.95)
+  let endArbBlock = currentArbBlock
+
+  async function getL1Block(forL2Block: number) {
+    const l2Block = Math.max(nitroGenesisBlock, forL2Block)
+    const { l1BlockNumber } = await arbitrumProvider.getBlock(l2Block)
+    return l1BlockNumber
+  }
+
+  // Binary search to find the starting Arbitrum block that corresponds to the L1 block number.
+  async function getL2StartBlock() {
+    let result
+    let start = startArbBlock
+    let end = endArbBlock
+
+    while (start <= end) {
+      // Calculate the midpoint of the current range.
+      const mid = start + Math.floor((end - start) / 2)
+
+      const l1Block = await getL1Block(mid)
+
+      // If the midpoint matches the target, we've found a match.
+      // Adjust the range to search for the first occurrence.
+      if (l1Block === targetL1BlockNumber) {
+        result = mid
+        end = mid - 1
+      } else if (l1Block < targetL1BlockNumber) {
+        // If the L1 block number is less than the target, adjust the range to the upper half.
+        start = mid + 1
+      } else {
+        // If the L1 block number is greater than the target, adjust the range to the lower half.
+        end = mid - 1
+      }
+    }
+
+    if (typeof result === 'undefined') {
+      throw new Error(`No L2 range found for L1 block: ${targetL1BlockNumber}`)
+    }
+
+    return BigNumber.from(result)
+  }
+
+  // Binary search to find the ending Arbitrum block that corresponds to the L1 block number.
+  async function getL2EndBlock() {
+    let result
+    let start = startArbBlock
+    let end = endArbBlock
+
+    while (start <= end) {
+      // Calculate the midpoint of the current range.
+      const mid = start + Math.floor((end - start) / 2)
+
+      const l1Block = await getL1Block(mid)
+
+      // If the midpoint matches the target, we've found a match.
+      // Adjust the range to search for the last occurrence.
+      if (l1Block === targetL1BlockNumber) {
+        result = mid
+        start = mid + 1
+      } else if (l1Block < targetL1BlockNumber) {
+        // If the L1 block number is less than the target, adjust the range to the upper half.
+        start = mid + 1
+      } else {
+        // If the L1 block number is greater than the target, adjust the range to the lower half.
+        end = mid - 1
+      }
+    }
```

Comment:
we now have binary search on block in at least 3 different place of the sdk codebase, maybe we can try to refactor it to reduce code duplication?